torch/distributions/wishart.py

```diff
@@ -0,0 +1,278 @@
+from torch.distributions.utils import lazy_property
+from torch.distributions.multivariate_normal import _precision_to_scale_tril
+
+
+_log_2 = math.log(2)
+
+
+def _mvdigamma(x: torch.Tensor, p: int) -> torch.Tensor:
+    assert x.gt((p - 1) / 2).all(), "Wrong domain for multivariate digamma function."
+    return torch.digamma(
+        x.unsqueeze(-1)
+        - torch.arange(p, dtype=x.dtype, device=x.device).div(2).expand(x.shape + (-1,))
+    ).sum(-1)
+
+class Wishart(ExponentialFamily):
+    r"""
+    Creates a Wishart distribution parameterized by a symmetric positive definite matrix :math:`\Sigma`,
+    or its Cholesky decomposition :math:`\mathbf{\Sigma} = \mathbf{L}\mathbf{L}^\top`
+
+    Example:
+        >>> m = Wishart(torch.eye(2), torch.Tensor([2]))
+        >>> m.sample()  #Wishart distributed with mean=`df * I` and
+                        #variance(x_ij)=`df` for i != j and variance(x_ij)=`2 * df` for i == j
+    Args:
+        covariance_matrix (Tensor): positive-definite covariance matrix
+        precision_matrix (Tensor): positive-definite precision matrix
+        scale_tril (Tensor): lower-triangular factor of covariance, with positive-valued diagonal
+        df (float or Tensor): real-valued parameter larger than the (dimension of Square matrix) - 1
+    Note:
+        Only one of :attr:`covariance_matrix` or :attr:`precision_matrix` or
+        :attr:`scale_tril` can be specified.
+        Using :attr:`scale_tril` will be more efficient: all computations internally
+        are based on :attr:`scale_tril`. If :attr:`covariance_matrix` or
+        :attr:`precision_matrix` is passed instead, it is only used to compute
+        the corresponding lower triangular matrices using a Cholesky decomposition.
+        'torch.distributions.LKJCholesky' is a restricted Wishart distribution.[1]
+
+    **References**
+
+    [1] `On equivalence of the LKJ distribution and the restricted Wishart distribution`,
+    Zhenxun Wang, Yunan Wu, Haitao Chu.
+    """
+    arg_constraints = {
+        'covariance_matrix': constraints.positive_definite,
+        'precision_matrix': constraints.positive_definite,
+        'scale_tril': constraints.lower_cholesky,
+        'df': constraints.greater_than(0),
+    }
+    support = constraints.positive_definite
+    has_rsample = True
+
+    def __init__(self,
+                 df: Union[torch.Tensor, Number],
+                 covariance_matrix: torch.Tensor = None,
+                 precision_matrix: torch.Tensor = None,
+                 scale_tril: torch.Tensor = None,
+                 validate_args=None):
+        assert (covariance_matrix is not None) + (scale_tril is not None) + (precision_matrix is not None) == 1, \
+            "Exactly one of covariance_matrix or precision_matrix or scale_tril may be specified."
+
+        param = next(p for p in (covariance_matrix, precision_matrix, scale_tril) if p is not None)
+
+        assert param.dim() > 1, \
```

Comment:
nit: Let us raise a `ValueError` here instead of an `assert` since this is not an internal check. Also, I just noticed that we don't raise an error if `df` is lower than `event_shape[-1]`, maybe also worth raising a ValueError for that?

```python
if df <= ndim - 1:
    raise Value(f"Value of df={df} expected to be greater than {ndim-1}.")
```

Reply:
Thanks for your comment. I will change the `assert` to `ValueError` in this cases.
I did not considered the part checking the valude of `df` either, since the constraint of `df` is set by `self.arg_constraints['df'] = constraints.greater_than(event_shape[-1] - 1)`. It seems better to add `ValueError` for `df` to check at class initialization. I will update the code.